### ISSUE TITLE
[Github] credentials to the private  repository within the devfile. 

### ISSUE THREAD

**Issue by pgirishibm** [eclipse-che/che#18963]
## Summary
Can a devfile take the credentials to github via a secret?

## Relevant information

The devfile for an eclipse che workspace is pointing to a private repository. But when a workspace is created using this devfile, we need to download the sample app from the private repository without having the user to enter the credentials. 

Using this, we can track the users who are creating workspaces using this devfile and would  be using the sample app.  The option to put the sample code into a public repo would make it difficult to track who are using the code in the repository and for what purpose. 

This is the section in the devfile which could do with an option to add the secret.

projects:
  - name: ipm-pivot
    source:
      location: '<private github repo>'
      type: git
      branch: master

**Comment by mmorhun** [comment]
cc @mshaposhnik @vinokurig

**Comment by mshaposhnik** [comment]
Currently, the only possible way to clone from private repos is described here:
https://www.eclipse.org/che/docs/che-7/end-user-guide/mounting-a-secret-as-a-file-or-an-environment-variable-into-a-workspace-container/#mounting-a-git-credential-store-into-a-workspace-container_che

**Comment by ericwill** [comment]
> Currently, the only possible way to clone from private repos is described here:
> https://www.eclipse.org/che/docs/che-7/end-user-guide/mounting-a-secret-as-a-file-or-an-environment-variable-into-a-workspace-container/#mounting-a-git-credential-store-into-a-workspace-container_che

Users will also be prompted to enter their credentials in the terminal, should they start a devfile that contains a private repository: https://github.com/eclipse/che/issues/16401

**Comment by skabashnyuk** [comment]
As a first step, we can make something similar to BItbucket personal access tokens. https://www.eclipse.org/che/docs/che-7/administration-guide/authenticating-users-3rd-party-services/#configuring_bitbucket_authentication_che

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: github-personal-access-token-secret
  labels:
    app.kubernetes.io/part-of: che.eclipse.org
    app.kubernetes.io/component: scm-personal-access-token
  annotations:
    che.eclipse.org/expired-after: '-1'
    che.eclipse.org/che-userid: '355d1ce5-990e-401e-9a8c-094bca10b5b3'
    che.eclipse.org/scm-userid: '2'
    che.eclipse.org/scm-username: 'user-foo'
data:
  token: TlRnNE1UazRORFl5TWpNeU9sMTI1aDFFT2dRbXBlTUYvbmhiLzNQUC9hT08=
```

**Comment by pgirishibm** [comment]
> Currently, the only possible way to clone from private repos is described here:
> https://www.eclipse.org/che/docs/che-7/end-user-guide/mounting-a-secret-as-a-file-or-an-environment-variable-into-a-workspace-container/#mounting-a-git-credential-store-into-a-workspace-container_che

This did work. But I see the user creating the workspace does get access to the personal access tokens within the developer container of the workspace. Is there a way to avoid that? Using that token the developer might be able to get to the other repos in the org right?

**Comment by skabashnyuk** [comment]
> Is there a way to avoid that? 

No


>Using that token the developer might be able to get to the other repos in the org right?

Yes. Do you see an issue with that?

**Comment by pgirishibm** [comment]
Thank you @skabashnyuk 

Well yes. don't want to open up the other repos to the user.  Can I provide an encrypted token and during workspace load decrypt it just for downloading?  So what the user sees in the workspace would be an encrypted token which cannot be used. 

So can I customize the download  steps of the sample app from the repo mentioned in the devfile?

**Comment by mshaposhnik** [comment]
@pgirishibm   Thats looks more like your SCM authorizarion issue, isn't it?  We suppose not only to read code from repository, but also to develop and commit it, so token should be present in usable form.   Gaining/rejecting access of the user to other repos is 
clearly not the consumer-s aim.

**Comment by skabashnyuk** [comment]
>Well yes. don't want to open up the other repos to the user. 

@pgirishibm why do you think that requires opening up the other repos to the user?

**Comment by pgirishibm** [comment]
> > Well yes. don't want to open up the other repos to the user.
> 
> @pgirishibm why do you think that requires opening up the other repos to the user?

@skabashnyuk Since the git Personal Access Token is available to the user as an environment variable, they can share it with others and the same repo can be cloned from outside Eclipse che as well right? Also using the same token other repos that the token has access to can be cloned?

**Comment by pgirishibm** [comment]
> @pgirishibm Thats looks more like your SCM authorizarion issue, isn't it? We suppose not only to read code from repository, but also to develop and commit it, so token should be present in usable form. Gaining/rejecting access of the user to other repos is
> clearly not the consumer-s aim.

Well, we are just providing sample code for the user to build and run. any changes they make to it, we want them to upload it to their own source code repositories and future access.  The same sample code would be accessed by multiple users, so cannot have all users edit it right?

**Comment by skabashnyuk** [comment]
>@skabashnyuk Since the git Personal Access Token is available to the user as an environment variable, they can share it with others and the same repo can be cloned from outside Eclipse che as well right

We could have some misunderstanding here. Let's try to clarify that.  Personal Access Token is individual to the user and should **NOT** be shared with others. It is set as k8s secret to the namespace that belongs to the user and not shared with others. Personal Access Token should not be pushed to the git in any of the forms. 

Are we on the same page here?

>Also using the same token other repos that the token has access to can be cloned?

Sorry I didn't get your question. Can you rephrase?

>Well, we are just providing sample code for the user to build and run. any changes they make to it, we want them to upload it to their own source code repositories and future access. The same sample code would be accessed by multiple users, so cannot have all users edit it right?

They can create a fork and when takes the fork and explore it with Eclipse Che. WDYT?

**Comment by gautam-borkar** [comment]
Can we provide read-only permissions for `/home/theia/.git-credentials` after the devfile has been loaded?
```
apiVersion: v1
kind: Secret
metadata:
  name: <secret-name>
  labels:
    app.kubernetes.io/part-of: che.eclipse.org
    app.kubernetes.io/component: workspace-secret
  annotations:
    che.eclipse.org/automount-workspace-secret: 'true'
    che.eclipse.org/mount-path: /home/theia/.git-credentials
    che.eclipse.org/mount-as: file
    che.eclipse.org/git-credential: 'true'
data:
  credentials: <base-64-encoded key>
```

**Comment by skabashnyuk** [comment]
>Can we provide read-only permissions for /home/theia/.git-credentials after the devfile has been loaded?

Are you talking about `git-credetials` or personal  `BItbucket personal access tokens`? Can you explain the use-case? I'm still can't understand the problem you are trying to fix.

**Comment by gautam-borkar** [comment]
> > Can we provide read-only permissions for /home/theia/.git-credentials after the devfile has been loaded?
> 
> Are you talking about `git-credetials` or personal `BItbucket personal access tokens`? Can you explain the use-case? I'm still can't understand the problem you are trying to fix.

I am talking about `git-credentials`.  

Our use case is as follow :- 
- Load a repo from our enterprise github into che
- after the repo is loaded we disable access to our enterprise repo

**Comment by skabashnyuk** [comment]
>after the repo is loaded we disable access to our enterprise repo

You can do that on the `enterprise github ` side. Just by removing the user's read permissions.

**Comment by gautam-borkar** [comment]
Thanks for the update @skabashnyuk !

Some additions to my use case :- 
- Some of the user accessing the che would be outside the organisation. 
- We will have one functional id for all the user outside the org, using which code will be cloned for them 
- After the code is cloned to their workspace, they should not have any access to the enterprise github.

**Comment by skabashnyuk** [comment]
>We will have one functional id for all the user outside the org, using which code will be cloned for them

That makes things complex. Have you considered creating an account per user on the Github side? Or making temporary repositories with unique names per user?

**Comment by gautam-borkar** [comment]
> > We will have one functional id for all the user outside the org, using which code will be cloned for them
> 
> That makes things complex. Have you considered creating an account per user on the Github side? Or making temporary repositories with unique names per user?

@skabashnyuk Thanks for the response. As of now we are planning to go with github public repo. So for now that should take care of the issue.

**Comment by cccs-eric** [comment]
This issue will most likely be resolved by [this](https://github.com/eclipse/che/issues/19966).

**Comment by mshaposhnik** [comment]
Yup, both OAuth abnd PAT authentication are supported, closing now.
